### PR TITLE
2020

### PR DESCRIPTION
**Pull request by MichaelViveros**
Picked Ruby to get better at writing more idiomatic code. I touched a bit of Ruby in my 3 years at PagerDuty (just minor changes to their Rails monolith, did a lot more Elixir) and I've been touching it more and more at OpsLevel (also a Rails monolith but one I work with a lot more). 

PS: I miss functional programming so my Advent of Code solutions will probably be unnecessarily functional 😛 .

### CODE REVIEW

File: 2020/2/valid_passwords.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+class Password
+  attr_accessor :policy_letter, :policy_constraint1, :policy_constraint2, :letter_counts
+
+  def initialize(policy_password_s)
+    regex = /([[:digit:]]+)\-([[:digit:]]+) ([[:alpha:]]): ([[:alpha:]]+)/
```
Comment (MichaelViveros) [inline 2020/2/valid_passwords.rb]: © @shruthilayaj

File: 2020/3/toboggan_trajectory.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+def trees_on_path(map)
+  rows_size = map.size
+  columns_size = map[0].size
+  # path is a series of slopes where each slope moves right 3 and down 1
+  # path ends when we reach last row, so there's rows_size slopes in total
+  rows = (1..rows_size - 1)
+  # columns repeat forever but we only get the initial set of columns
+  # so we have to take modulo
+  columns = (1..rows_size - 1).map { |c| c * 3 % columns_size }
+  # [[1, 3], [2, 6], ...]
```
Comment (shruthilayaj) [inline 2020/3/toboggan_trajectory.rb]: Cool, you make a list of coordinates you need to loop through

Comment (MichaelViveros) [reply 2020/3/toboggan_trajectory.rb]: yup! The `zip` method came up clutch for this (never really used it before)

File: 2020/6/customs_forms.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# @param group_answers [Array<String>] the group's answers
+# @return [Integer] the count of unique answers across all group
+# members
```
Comment (MichaelViveros) [inline 2020/6/customs_forms.rb]: Adding [YARD](https://yardoc.org/) docs for readability and for enhanced type support in VSCode (via vscode-solargraph).

Ex. auto-complete shows `Array` methods:
Screen Shot 2020-12-06 at 11 06 23 AM

File: 2020/5/binary_seats.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    binary_move(start_i, end_i, char, 'L')
+  end.first
+  row * 8 + column
+end
+
+# @param seat [String] encoded seat number like 'FBFFBBFRLR'
+# @return [Integer] decoded seat number
+# decodes row by converting encoded row to binary number
+# does same thing for column
+def binary_partition2(seat)
```
Comment (MichaelViveros) [inline 2020/5/binary_seats.rb]: After looking at some other people's code, I learnt that the encoded seat number is just a binary number. Encountering a `B` means you move `2^i` rows back, `F` means you stay where you are. Here's an example calculation:
```
FBFFBBF, 0
0100110, 0
|           0 + 0 * 2^6 = 0
0100110, 0
 |          0 + 1 * 2^5 = 32
0100110, 32
  |         32 + 0 * 2^4 = 32
0100110, 32
   |        32 + 0 * 2^3 = 32
0100110, 32
    |       32 + 1 * 2^2 = 36
0100110, 36
     |      40 + 1 * 2^1 = 38
0100110, 38
      |     44 + 0 * 2^2 = 38
```

**Pull request merged**